### ISSUE TITLE
citadel issues longer certificate ttl than the ttl of issuing intermediate

### ISSUE THREAD

**Issue by juliantaylor** [istio/istio#18990]
**Bug description**
If the `--workload-cert-ttl` of citadel is longer than the ttl of the intermediate certificate the that is used to issue the certificates. Your workload certificates will become invalid when the intermediate expires and citadel will not refresh them as the certificate itself is not yet expired.

[ x ] Security

**Expected behavior**
One would expect that the maximum workload certificate ttl is bounded by the ttl of the intermediate certificate used for issuing. Currently one has to ensure that the citadel commandline argument `--workload-cert-ttl` is lower than the intermediate ttl for the renewal to work correctly.

**Steps to reproduce the bug**

Setup istio citadel with a intermediate certificate with a short ttl. Our configuration is:
```
        - --self-signed-ca=false
        - --signing-cert=/etc/cacerts/ca-cert.pem
        - --signing-key=/etc/cacerts/ca-key.pem
        - --root-cert=/etc/cacerts/root-cert.pem
        - --cert-chain=/etc/cacerts/cert-chain.pem
        - --workload-cert-ttl=2160h # the default
```
with the intermediate certificate issued from hashicorp vault, the intermediate must have a ttl smaller than the citadel setting.
Check expiry of e.g. a new mixer secret `istio.istio-mixer-service-account`
You will see the expiry is 3 month in the future regardless of the intermediate ttl.

Wait until the intermediate expiry (and update citadel with a new intermediate). Soon mixer, telemetry, pilot etc will stop working with following errors:

```
2019-11-15T14:35:45.436786Z	error	mcp	Failed to create a new MCP sink stream: rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = "transport: authentication handshake failed: x509: certificate has expired or is not yet valid"
```

Note citadel does have the new intermediate certificate and can issue valid certificates, it just doesn't due to the too long workload certificate ttl.

Citadel refresh itself works fine, this was tested in the same version by setting the `--workload-cert-ttl` to a few minutes.

**Version (include the output of `istioctl version --remote` and `kubectl version` and `helm version` if you used Helm)**

```
client version: 1.3.4
control plane version: 1.3.4
```

**How was Istio installed?**
helm charts

**Comment by incfly** [comment]
What you describe is reasonable from PKI point of view, but the intermediate certificate is provided by external CA, in this case, Citadel can't _guess_ what's the TTL workload cert should be. Shorter, how much? 50%, or 99%?

Also I believe the workload cetificate TTL is configurable, which is the right configuration knob advanced user in this case should change correspondingly.

Reasonable handling

- Add some log warning message during startup time.
- Ensure workload TTL is configurable.

I'm closing this, but feel free to reopen or send a PR to address above items.

**Comment by juliantaylor** [comment]
Citadel does not need to guess what the workload ttl needs to be, its maximum ttl must just not be larger than the ttl of the intermediate else it is creating an invalid certificate (at least for istios on certificate verification functions which consider the intermediate ttl). That way citadels own certificate renewal functions work as intended.
Of course the intermediate needs to be renewed by an external component in time, but that should be obvious when deploying citadel with non infinite ttl intermediates anyway.

The workload ttl is configurable, what we have done is set the workload ttl of citadel to less than the renewal fraction (we renew the intermediate after 1/3 of its ttl is expired, so the max workload ttl is 1/3 of the intermediate ttl plus some buffer for failures).
But this couples two independent components which do not need to be coupled, the intermediate certificate creation and the citadel configuration. citadel has all the information it needs to configure itself correctly without extra input.

What citadel could do is warn if the remaining ttl is very low which may be inefficient but works. A warning that it was wrongly configured and will break on future is not so reliable as easily overlooked.
Citadel failing to start if the remaining ttl is too low might be an acceptable option too.

With istios current settings all deployments using an externally provided intermediate certificate **will** break completely by default when the intermediate expires. All components using mtls will stop working.
This is a situation which should be avoided by sane default behaviour of citadel.

**Comment by juliantaylor** [comment]
I do not seem to be able to reopen the issue but I do not consider this closed without at least a documentation update noting this major risk with externally provided certificates, e.g. here https://istio.io/docs/tasks/security/citadel-config/plugin-ca-cert/

Does citadel handle this when using self signed certificates?
According to https://istio.io/docs/ops/security/root-transition/ they also have a ttl, so the same situation should occur when the workload ttl exceeds the root ca ttl if the mtls components do consider the root ca expiry in validating the certificate like they do for intermediates.

**Comment by incfly** [comment]
> The workload ttl is configurable, what we have done is set the workload ttl of citadel to less than the renewal fraction (we renew the intermediate after 1/3 of its ttl is expired, so the max workload ttl is 1/3 of the intermediate ttl plus some buffer for failures).

Okay, this might be reasonable defaults for the self-provided intermediate CA case. Plus customer can also override with a more explicit `--workload-cert-ttl`.

>  A warning that it was wrongly configured and will break on future is not so reliable as easily overlooked. Citadel failing to start if the remaining ttl is too low might be an acceptable option too.

Good reliability overlook argument.

I feel this is something generally right to do, I'll delegate to  @myidpt who has more input on overall Citadel configuration knobs.

**Comment by juliantaylor** [comment]
unless this issue was fixed in 1.5 which I did not check, please reopen.
with this issue any istio deployment will completely break as soon as the intermediates expire unless undocumented workarounds are applied.

**Comment by howardjohn** [comment]
Fixed by https://github.com/istio/istio/pull/25158
